java/src/main/java/com/google/oak/transport/GrpcStreamingTransport.java

```diff
@@ -0,0 +1,138 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package com.google.oak.transport;
+
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+import com.google.oak.transport.EvidenceProvider;
+import com.google.oak.transport.Transport;
+import com.google.oak.util.Result;
+import com.google.protobuf.ByteString;
+import io.grpc.ManagedChannel;
+import io.grpc.Status;
+import io.grpc.stub.StreamObserver;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.function.Function;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import oak.session.noninteractive.v1.AttestationBundle;
+import oak.session.noninteractive.v1.GetPublicKeyRequest;
+import oak.session.noninteractive.v1.GetPublicKeyResponse;
+import oak.session.noninteractive.v1.InvokeRequest;
+import oak.session.noninteractive.v1.InvokeResponse;
+import oak.session.noninteractive.v1.RequestWrapper;
+import oak.session.noninteractive.v1.ResponseWrapper;
+
+public class GrpcStreamingTransport implements EvidenceProvider, Transport {
+  private static final Logger logger = Logger.getLogger(GrpcStreamingTransport.class.getName());
+
+  private BlockingQueue<ResponseWrapper> messageQueue;
+  private StreamObserver<RequestWrapper> requestObserver;
+
+  /**
+   * Creates an instance of {@code GrpcStreamingTransport}.
+   *
+   * @param stream a method reference to a gRPC client streaming method with the appropriate request
+   * and response types.
+   */
+  public GrpcStreamingTransport(
+      Function<StreamObserver<ResponseWrapper>, StreamObserver<RequestWrapper>> stream) {
+    // We expect to receive a single response message for each request.
+    this.messageQueue = new ArrayBlockingQueue<>(1);
+    StreamObserver<ResponseWrapper> responseObserver = new StreamObserver<ResponseWrapper>() {
+      @Override
+      public void onNext(ResponseWrapper response) {
+        try {
+          messageQueue.put(response);
+        } catch (Exception e) {
+          if (e instanceof InterruptedException) {
+            Thread.currentThread().interrupt();
+          }
+          logger.log(Level.WARNING, "Couldn't send server response to the message queue: " + e);
+        }
+      }
+
+      @Override
+      public void onError(Throwable t) {
+        Status status = Status.fromThrowable(t);
+        logger.log(Level.WARNING, "Couldn't receive response: " + status);
+      }
+
+      @Override
+      public void onCompleted() {
+        logger.log(Level.FINE, "response message queue completed");
+      }
+    };
+    this.requestObserver = stream.apply(responseObserver);
+  }
+
+  /**
+   * Returns evidence about the trustworthiness of a remote server.
+   *
+   * @return {@code AttestationBundle} wrapped in a {@code Result}
```

Comment:
Document also the error case.

Is not it better to return a string error message, seeing that the only error that is returned is an interrupted exception?